codegen/src/main/java/software/amazon/awssdk/codegen/poet/auth/scheme/EndpointBasedAuthSchemeProviderSpec.java

```diff
@@ -0,0 +1,181 @@
+        spec.addStatement("return MODELED_RESOLVER.resolveAuthScheme(params)");
+        spec.endControlFlow();
+
+
+        spec.addStatement("$T options = new $T<>()", ParameterizedTypeName.get(List.class, AuthSchemeOption.class),
+                          TypeName.get(ArrayList.class));
+        spec.beginControlFlow("for ($T authScheme : authSchemes)", EndpointAuthScheme.class);
+        addAuthSchemeSwitch(spec);
+        spec.endControlFlow();
+        return spec.addStatement("return $T.unmodifiableList(options)", Collections.class)
+                   .build();
+    }
+
+    private void addAuthSchemeSwitch(MethodSpec.Builder spec) {
+        spec.addStatement("$T name = authScheme.name()", String.class);
+        spec.beginControlFlow("switch(name)");
+
+        spec.addCode("case $S:", "sigv4");
+        spec.addStatement("$T sigv4AuthScheme = $T.isInstanceOf($T.class, authScheme, $S, authScheme.getClass().getName())",
+                          SigV4AuthScheme.class, Validate.class, SigV4AuthScheme.class,
+                          "Expecting auth scheme of class SigV4AuthScheme, got instead object of class %s");
+
+        spec.addCode("options.add($T.builder().schemeId($S)", AuthSchemeOption.class, "aws.auth#sigv4")
+            .addCode(".putSignerProperty($T.SERVICE_SIGNING_NAME, sigv4AuthScheme.signingName())", AwsV4HttpSigner.class)
+            .addCode(".putSignerProperty($T.REGION_NAME, sigv4AuthScheme.signingRegion())", AwsV4HttpSigner.class)
+            .addCode(".putSignerProperty($T.DOUBLE_URL_ENCODE, negateOrDefault(sigv4AuthScheme.disableDoubleEncoding()))",
+                     AwsV4HttpSigner.class)
+            .addCode(".build());");
+        spec.addStatement("break");
+
+        // SigV4a -- Not yet implemented, we throw UnsupportedOperationException for now.
+        spec.addCode("case $S:", "sigv4a");
+        spec.addStatement("throw new $T($S)", UnsupportedOperationException.class, "SigV4a is not yet supported.");
+
+
+        spec.addCode("default:")
+            .addStatement("throw new $T($S + name)", IllegalArgumentException.class, "Unknown auth scheme: ");
+        spec.endControlFlow();
+    }
+
+    private MethodSpec addUtilNegateOrDefault() {
+        MethodSpec.Builder spec = MethodSpec.methodBuilder("negateOrDefault")
+                                            .addModifiers(Modifier.PRIVATE, Modifier.STATIC)
+                                            .returns(Boolean.class)
+                                            .addParameter(Boolean.class, "value");
+        return spec.beginControlFlow("if (value != null)")
+                   .addStatement("return !value")
+                   .endControlFlow()
+                   .addStatement("return value")
+                   .build();
+    }
+
+    private Map<String, ParameterModel> parameters() {
+        return authSchemeSpecUtils.intermediateModel().getEndpointRuleSetModel().getParameters();
```

Comment:
refactoring: seeing that you added `intermediateModel()` to `AuthSchemeSpecUtils` for calling it here, and also that there is this `parameters()` method in multiple places, I think it seems better to have `EndpointRulesSpecUtils.parameters()` (as it's related to endpoints) and have any existing logic that has its own `parameters()` call EndpointRulesSpecUtils instead - if it has logic related to endpoints, seems reasonable for it to have a reference to EndpointRulesSpecUtils.